### PR TITLE
Improvements to ContextualService

### LINKED ISSUE #1546: Thread safety and Context Calculators
According to the SpongeDocs, "there are a few parts of Minecraft [and the API] that you can work with asynchronously". One of the given examples is "**permissions handling**".

It was also stated by zml in a permissions implementation document that:
>The plan was to require permissions plugins to be thread-safe

That was certainly also my understanding until now. 

However, it was perhaps overlooked that the permissions API also depends on other systems within Sponge, namely contexts.

Fundamentally, if the permissions API is deemed to be thread safe, then all [`ContextCalculator`](https://github.com/SpongePowered/SpongeAPI/blob/bleeding/src/main/java/org/spongepowered/api/service/context/ContextCalculator.java) implementations must be thread safe too.

The call tree for most permissions checks will look something like this:
1. Plugin checks for permission using Subject#hasPermission (possibly making this call from an async thread)
2. The permission plugin queries the ContextCalculators it [has registered to it](https://github.com/SpongePowered/SpongeAPI/blob/bleeding/src/main/java/org/spongepowered/api/service/context/ContextualService.java#L36) in order to determine the subjects "current context".
3. Permission plugin does whatever it needs to do to work out what the result of the check should be.

Obviously, in this process, it's possible that the calculators are called async. Therefore, the calculators should also be thread safe, and can therefore only access a very limited range of data and API functions.

Options to resolve this (unless anyone else has any ideas):

1. Declare that the permissions API is absolutely **not thread safe**, and that ALL permissions checks, lookup operations, or anything which involves contexts MUST be done on the main server thread.

2. Require that ContextCalculators are thread safe, and do not access any non-thread safe aspect of Sponge or the server.

I personally lean towards option no 2. It should be relatively straight forward for implementations to cache the data they need, so it can be accessed.

For example, if I wanted a calculator which simply returned the name of the block a player was standing (a pretty standard request for a context system, I would say), then it would require that I cache that data in a Concurrent Map, and then return values from there when the calculator is called. I'd then update this cache map on the move event.

Either way, it needs to be addressed & then documented.

I'd like to know what other people think, or if a better solution can be found. :)

### PR DESCRIPTION
**Pull request by lucko**
Fixes #1546 - still WIP for now

The contexts system in particular can be difficult for developers new to Sponge (esp. sponge permissions) to understand - the hope is that some better javadocs will help a bit with that.

If anyone has anything they think could be clarified/improved/explained better please say. :)

### CODE REVIEW

**Comment by Aaron1011** [comment]
@lucko Are you planning to make any breaking changes? If not, this looks like like it could target `stable-7`

**Comment by lucko** [comment]
Not sure yet, there *probably* won't be any, but I thought it made sense to target bleeding for now - can always cherry pick it back.

The only thing I've considered so far is changing ContextCalculator's

```java
void accumulateContexts(T target, Set<Context> accumulator)
```

to 

```java
void accumulateContexts(T target, Consumer<Context> accumulator)
```

reasoning being that calculators shouldn't be relying on the state of the Set, and should only be using `add` and `addAll` - not sure it's worth the break though 🤷‍♂️

**Comment by lucko** [comment]
This is ready now.

* Mostly changes to javadocs
* No breaking changes, only one new static method added, so this can be backported to stable-7
* Two methods in `Context` were deprecated - reasoning explained in the doc. They can be removed in the next API release.


@bloodmc @Aaron1011 or anyone else kind enough to, would you mind giving the doc changes a quick read over to make sure they make sense. I'm finding it quite hard to put a lot of the concepts into words - it's almost too abstract in places to explain. :)

**Comment by gabizou** [comment]
Want to update the PR to target `stable-7` maybe?

File: src/main/java/org/spongepowered/api/service/context/ContextCalculator.java [hunk 1]
```diff
@@ -60,5 +141,9 @@
      *         and is applicable to the given {@link Contextual}. Otherwise
      *         false.
      */
-    boolean matches(Context context, T calculable);
+    default boolean matches(Context context, T calculable) {
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/service/context/ContextCalculator.java]: I would rename `calculable` to `target` to match the other method

**Review by ST-DDT — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/api/service/context/ContextCalculator.java (hunk 1 above)
Comment (lucko) [reply src/main/java/org/spongepowered/api/service/context/ContextCalculator.java]: Done, thanks

**Comment by lucko** [comment]
Updated (not so successfully, sorry for the review request pings 😢) to target `stable-7`.

Also addressed ST-DDT's suggestion. This is ready to be merged now, unless anyone has any final suggestions / comments.

**Review by ST-DDT — APPROVED** [review]

**Pull request merged**